packages/client/lib/execution/vmexecution.ts

```diff
@@ -302,82 +302,105 @@ export class VMExecution extends Execution {
-                } time=${diffSec}secs`
-                this.config.logger.warn(msg)
+              let skipBlockValidation = false
+              if (this.config.execCommon.consensusType() === ConsensusType.ProofOfAuthority) {
+                // Block validation is redundant here and leads to consistency problems
+                // on PoA clique along blockchain-including validation checks
+                // (signer states might have moved on when sync is ahead)
+                skipBlockValidation = true
               }
 
-              void this.receiptsManager?.saveReceipts(block, result.receipts)
-            })
-            txCounter += block.transactions.length
-            // set as new head block
-            headBlock = block
-            parentState = block.header.stateRoot
-          } catch (error: any) {
+              await this.runWithLock<void>(async () => {
+                // we are skipping header validation because the block has been picked from the
+                // blockchain and header should have already been validated while putBlock
+                if (!this.started) {
+                  throw Error('Execution stopped')
+                }
+                const beforeTS = Date.now()
+                this.stats(this.vm)
+                const result = await this.vm.runBlock({
+                  block,
+                  root: parentState,
+                  clearCache: reorg ? true : false,
+                  skipBlockValidation,
+                  skipHeaderValidation: true,
+                })
+                const afterTS = Date.now()
+                const diffSec = Math.round((afterTS - beforeTS) / 1000)
+
+                if (diffSec > this.MAX_TOLERATED_BLOCK_TIME) {
+                  const msg = `Slow block execution for block num=${
+                    block.header.number
+                  } hash=0x${bytesToHex(block.hash())} txs=${block.transactions.length} gasUsed=${
+                    result.gasUsed
+                  } time=${diffSec}secs`
+                  this.config.logger.warn(msg)
+                }
+
+                void this.receiptsManager?.saveReceipts(block, result.receipts)
+              })
+
+              txCounter += block.transactions.length
+              // set as new head block
+              headBlock = block
+              parentState = block.header.stateRoot
+            } catch (error: any) {
+              // Store error block and throw which will make iterator stop, exit and save
+              // last successfully executed head as vmHead
+              errorBlock = block
+              throw error
+            }
+          },
+          this.config.numBlocksPerIteration,
+          // release lock on this callback so other blockchain ops can happen while this block is being executed
+          true
+        )
+        // Ensure to catch and not throw as this would lead to unCaughtException with process exit
+        .catch(async (error) => {
```

Comment:
If we catch the error here, what error is thrown that tells the client to exit?

Reply:
Ah, right, this is just handling an error that occurs when the client is already in the process of exiting.  Makes sense.